pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1214,4 +1243,92 @@ private int getMaxNumPartitionsPerInstance(InstancePartitions instancePartitions
+          if (peerSegmentURIs.isEmpty()) {
+            _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.NUMBER_LLC_SEGMENTS_DEEP_STORE_UPLOAD_FIX_ERROR, 1L);
+            LOGGER.error("Failed to upload segment {} to segment store because no online replica is found", segmentName);
+            continue;
+          }
+
+          // Randomly ask one server to upload
+          URI uri = peerSegmentURIs.get(RANDOM.nextInt(peerSegmentURIs.size()));
+          String serverUploadRequestUrl = StringUtil.join("/", uri.toString(), "upload");
+          LOGGER.info("Ask server to upload llc segment {} to segment store by this path: {}", segmentName, serverUploadRequestUrl);
+          String segmentDownloadUrl = uploadLLCSegmentByServer(serverUploadRequestUrl);
+
+          // Update the segment ZK metadata to include segment download url
+          if (segmentDownloadUrl.isEmpty()) {
+            _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.NUMBER_LLC_SEGMENTS_DEEP_STORE_UPLOAD_FIX_ERROR, 1L);
+            LOGGER.error("Failed to upload segment {} to segment store: no segment download url is returned from server.", segmentName);
+            continue;
+          }
+          segmentZKMetadata.setDownloadUrl(segmentDownloadUrl);
+          persistSegmentZKMetadata(realtimeTableName, segmentZKMetadata, -1);
+          LOGGER.info("Successfully uploaded llc segment {} to segment store with download url: {}", segmentName, segmentDownloadUrl);
+        } catch (Exception e) {
+          _controllerMetrics.addMeteredTableValue(realtimeTableName, ControllerMeter.NUMBER_LLC_SEGMENTS_DEEP_STORE_UPLOAD_FIX_ERROR, 1L);
+          LOGGER.error("Failed to upload segment {} to segment store", segmentName, e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Helper function to ask server to upload llc segment to segment store
+   */
+  private String uploadLLCSegmentByServer(String serverUploadRequestUrl)
+      throws URISyntaxException, IOException, HttpErrorStatusException {
+    HttpUriRequest request = buildServerUploadRequest(serverUploadRequestUrl);
+    try (CloseableHttpResponse response = _httpClient.execute(request)) {
+      int statusCode = response.getStatusLine().getStatusCode();
+      String responseStr = response.getEntity() == null ? "" : EntityUtils.toString(response.getEntity());
+      if (statusCode >= 300) {
+        StringBuilder errorMsgBuilder = new StringBuilder(String.format("Failed to ask server to upload LLC segment to segment store by url: %s. ", serverUploadRequestUrl));
+        if (!responseStr.isEmpty()) {
+          errorMsgBuilder.append(responseStr);
+        }
+        throw new HttpErrorStatusException(errorMsgBuilder.toString(), statusCode);
+      }
+
+      return responseStr;
+    }
+  }
+
+  @VisibleForTesting
+  HttpUriRequest buildServerUploadRequest(String serverUploadRequestUrl)
```

Comment:
Move this method to FileUploadDownloadClient.

Reply:
I made the suggestion to move this method here because this method sends request to server, while FileUploadDownloadClient sends request to controller. But seems we should leverage the `SSLContext` within the FileUploadDownloadClient for security

Reply:
Moved back to FileUploadDownloadClient

Reply:
@mcvsubbu FYI.